hnn_core/network.py

```diff
@@ -245,53 +246,54 @@ def pick_connection(net, src_gids=None, target_gids=None,
 
     # Create lookup dictionaries
-    src_dict, target_dict = dict(), dict()
-    loc_dict, receptor_dict = dict(), dict()
+    src_dict, target_dict = defaultdict(list), defaultdict(list)
+    loc_dict, receptor_dict = defaultdict(list), defaultdict(list)
     for conn_idx, conn in enumerate(net.connectivity):
         # Store connections matching each src_gid
         for src_gid in conn['src_gids']:
-            if src_gid in src_dict:
-                src_dict[src_gid].append(conn_idx)
-            else:
-                src_dict[src_gid] = [conn_idx]
+            src_dict[src_gid].append(conn_idx)
+
         # Store connections matching each target_gid
         for target_gid in conn['target_gids']:
-            if target_gid in target_dict:
-                target_dict[target_gid].append(conn_idx)
-            else:
-                target_dict[target_gid] = [conn_idx]
+            target_dict[target_gid].append(conn_idx)
+
         # Store connections matching each location
-        if conn['loc'] in loc_dict:
-            loc_dict[conn['loc']].append(conn_idx)
-        else:
-            loc_dict[conn['loc']] = [conn_idx]
+        loc_dict[conn['loc']].append(conn_idx)
+
         # Store connections matching each receptor
-        if conn['receptor'] in receptor_dict:
-            receptor_dict[conn['receptor']].append(conn_idx)
-        else:
-            receptor_dict[conn['receptor']] = [conn_idx]
+        receptor_dict[conn['receptor']].append(conn_idx)
 
     # Look up conn indices that match search terms and add to set.
     conn_set = set()
-    search_pairs = [(src_gids, src_dict), (target_gids, target_dict),
-                    (loc, loc_dict), (receptor, receptor_dict)]
-    for search_terms, search_dict in search_pairs:
-        inner_set = set()
-        # Union of indices which match inputs for single parameter
-        for term in search_terms:
-            inner_set = inner_set.union(search_dict.get(term, list()))
-        # Intersection across parameters
-        if conn_set and inner_set:
-            conn_set = conn_set.intersection(inner_set)
-            # If at any point there's no matching elements, return empty
-            if not conn_set:
-                return list()
-        else:
-            conn_set = conn_set.union(inner_set)
+    search_pairs = [(src_gids, src_gids_checked, src_dict),
+                    (target_gids, target_gids_checked, target_dict),
+                    (loc, loc_list, loc_dict),
+                    (receptor, receptor_list, receptor_dict),
+                    ]
+    for arg_input, search_terms, search_dict in search_pairs:
+        if arg_input is not None:
+            inner_set = set()
+            # Union of indices which match inputs for single parameter
+            for term in search_terms:
+                inner_set = inner_set.union(search_dict.get(term, list()))
+
+            # Empty search
+            if not inner_set:
+                return []
```

Comment:
`list()` is preferable than `[]` and `dict` is preferable than `{}` ... `[]` and `{}` are sometimes hard to tell apart

Reply:
It's actually faster to initialize with just brackets or curly brackets. But if you prefer the readability of the function call I can change it. 

 https://www.quora.com/Why-is-faster-than-list-Python-performance-list-initialization-literal-and-development#:~:text=%5B%20%5D%20declares%20an%20empty%20list.,from%20whatever%20its%20argument%20is.

Reply:
Interesting. I'd keep it list for consistency for now. If you'd like to make a repo-wide change in a separate PR, I'd be curious if there are any speed  improvements

Reply:
Probably not a noticeable difference, haha.